### PR TITLE
feat(orc8r): FEG_RELAY update to redirect sessiond request to N7_Proxy

### PR DESCRIPTION
**Pull request by Abhishek-R15**
## Summary

FEG_RELAY code changes to redirect sessiond request to N7_PROXY based on the RAT-TYPE 

## Additional Information

 Signed-off-by: Abhishek-R15 <abhishek.r@wavelabs.ai>

### CODE REVIEW

**Review by uri200 — CHANGES_REQUESTED** [review]
you may need to add an entry [here](
https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Amagma%2Fmagma.git&branch=master&file=feg%2Fcloud%2Fconfigs%2Fservice_registry.yml&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.2.2&start_row=27&start_col=9&end_row=27&end_col=9) so feg can forward this properly


Check [here](https://sourcegraph.com/-/editor?remote_url=git%40github.com%3Amagma%2Fmagma.git&branch=master&file=feg%2Fgateway%2Fregistry%2Flocal_registry.go&editor=JetBrains&version=v1.2.1&utm_product_name=IntelliJ+IDEA&utm_product_version=2021.2.2&start_row=34&start_col=37&end_row=34&end_col=37) too to see how other services piping is done

File: feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/relay_router.go [hunk 1]
```diff
@@ -29,6 +29,7 @@ const (
 	FegSessionProxy gateway_registry.GwServiceType = "session_proxy"
 	FegHello        gateway_registry.GwServiceType = "feg_hello"
 	FegSwxProxy     gateway_registry.GwServiceType = "swx_proxy"
+	FegN7Proxy      gateway_registry.GwServiceType = "n7_proxy"
```
Comment (uri200) [inline feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/relay_router.go:32]: shoudn;t this be `N7_N40_PROXY`

File: feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/session_proxy_relay.go [hunk 2]
```diff
@@ -76,7 +76,7 @@ func (s *RelayRouter) UpdateSession(
 	// send a separate Update request for each unique PLMN ID
 	for plmnid, req := range reqMap {
 		go func(plmnid string, req *protos.UpdateSessionRequest) {
-			client, ctx, cancel, err := s.getSessionProxyClient(ctx, plmnid)
+			client, ctx, cancel, err := s.getProxyClient(ctx, req.Updates[0].CommonContext.RatType, plmnid)
```
Comment (uri200) [inline feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/session_proxy_relay.go:79]: can't updates come for different types of  rats on the same report?

File: feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/session_proxy_relay.go [hunk 3]
```diff
@@ -129,6 +129,26 @@ func (s *RelayRouter) getSessionProxyClient(
...
+//Service selection based on the RAT Type
+func (s *RelayRouter) getProxyClient(
+	ctx context.Context, r protos.RATType, imsi string) (protos.CentralSessionControllerClient, context.Context, context.CancelFunc, error) {
+	if r == protos.RATType_TGPP_NR {
+		return s.getN7ProxyClient(ctx, imsi)
+	} else {
+		return s.getSessionProxyClient(ctx, imsi)
+	}
+}
+
```
Comment (uri200) [inline feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/session_proxy_relay.go:151]: place this functions at the end

File: feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/session_proxy_relay.go [hunk 4]
```diff
@@ -129,6 +129,26 @@ func (s *RelayRouter) getSessionProxyClient(
...
+	c context.Context, imsi string) (protos.CentralSessionControllerClient, context.Context, context.CancelFunc, error) {
+	conn, ctx, cancel, err := s.GetFegServiceConnection(c, imsi, FegN7Proxy)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	return protos.NewCentralSessionControllerClient(conn), ctx, cancel, nil
+}
+
+//Service selection based on the RAT Type
+func (s *RelayRouter) getProxyClient(
```
Comment (uri200) [inline feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/session_proxy_relay.go:143]: move this function before the other two. This way the reader can read incrementally

File: feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/session_proxy_relay.go [hunk 5]
```diff
@@ -129,6 +129,26 @@ func (s *RelayRouter) getSessionProxyClient(
...
+	ctx context.Context, r protos.RATType, imsi string) (protos.CentralSessionControllerClient, context.Context, context.CancelFunc, error) {
+	if r == protos.RATType_TGPP_NR {
+		return s.getN7ProxyClient(ctx, imsi)
+	} else {
+		return s.getSessionProxyClient(ctx, imsi)
+	}
```
Comment (uri200) [inline feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/session_proxy_relay.go:149]: use a switch, this way you can have a default to `getSessionProxyClient` and it is clearer

File: feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/session_proxy_relay.go [hunk 6]
```diff
@@ -129,6 +129,26 @@ func (s *RelayRouter) getSessionProxyClient(
 	return protos.NewCentralSessionControllerClient(conn), ctx, cancel, nil
 }
 
+//N7_proxy service for 5G
+func (s *RelayRouter) getN7ProxyClient(
```
Comment (uri200) [inline feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/session_proxy_relay.go:133]: `getN7N40ProxyClient`

Maybe we can rename the whole service to N740... .but that may be confusing. We can think about it

File: feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/session_proxy_relay.go [hunk 7]
```diff
@@ -129,6 +129,26 @@ func (s *RelayRouter) getSessionProxyClient(
 	return protos.NewCentralSessionControllerClient(conn), ctx, cancel, nil
 }
 
+//N7_proxy service for 5G
```
Comment (uri200) [inline feg/cloud/go/services/feg_relay/gw_to_feg_relay/servicers/session_proxy_relay.go:132]: i think you do not need different functions for each client. They are in fact only check if they are error, for you to check again on the main function

```
func (s *RelayRouter) getProxyClientctx context.Context, r protos.RATType, imsi string) (protos.CentralSessionControllerClient, context.Context, context.CancelFunc, error) {
	conn, ctx, cancel, err := s.GetFegServiceConnection(c, imsi, getProxyType(r))
	if err != nil {
		return nil, nil, nil, err
	}
	return protos.NewCentralSessionControllerClient(conn), ctx, cancel, nil

func getProxyType(r protos.RATType){
	switch(r):{
	case protos.RATType_TGPP_NR:
		return FegN7Proxy
	default:
		return FegSessionProxy
		
	}
}



```

**Review by uri200 — CHANGES_REQUESTED** [review]

**Comment by anusha-konda** [comment]
@uri200 , all the comments provided here will be addressed in PR #11046 (in Draft, will open for review). Hence closing this PR.

**Pull request closed without merging**